### PR TITLE
core: replace and remove DEBUGF macro and replace __FILE__ by RIOT_FILE_RELATIVE

### LINKED ISSUE #2699: DEBUGF does not compile if ENABLE_DEBUG == 1 and DEVELHELP == 0
`DEBUGF` requires access to the thread name inside a `tcb_t` struct. However, the member variable `name` is conditionally added only if DEVELHELP is true.

DEBUGF is not very widely used, mostly only inside the RPL and 6lowpan code, and I realize that the intention might have been to only use it when DEVELHELP is true, but I don't think it should not cause cryptic compilation errors when used improperly.

At least four (three) possible solutions: 
 - Change the condition to add the name to `tcb_t` to `#if ENABLE_DEBUG || DEVELHELP`
 - Change `DEBUGF` to only be available when `#if DEVELHELP`, and update the documentation accordingly.
 - Change the documentation and definition of `DEBUGF` so that it is clear that it should only be used when DEVELHELP is set.
 - Change `DEBUGF` to only print a name if it is available, e.g.:

```c
#if DEVELHELP
#define DEBUGF(...) \
    do { \
        DEBUG_PRINT("DEBUG(%s): %s:%d in %s: ", \
                sched_active_thread ? sched_active_thread->name : "NO THREAD", \
                __FILE__, __LINE__, DEBUG_FUNC); \
        DEBUG_PRINT(__VA_ARGS__); \
    } while (0)
#else
#define DEBUGF(...) \
    do { \
        DEBUG_PRINT("DEBUG: %s:%d in %s: ", __FILE__, __LINE__, DEBUG_FUNC); \
        DEBUG_PRINT(__VA_ARGS__); \
    } while (0)
#endif
```

### PR DESCRIPTION
**Pull request by OlegHahm**
fixes #2699.

Alternatively, we could just drop `DEBUGF` since it's almost not used at all.

### CODE REVIEW

**Comment by OlegHahm** [comment]
Anyone with any opinion?

**Comment by kaspar030** [comment]
I don't like that it will "... in N/A", PID would be nice. But I don't really care. Merge as-is?

**Comment by OlegHahm** [comment]
Or remove `DEBUGF` completely?

**Comment by OlegHahm** [comment]
removed `DEBUGF` macro completely.

File: core/include/debug.h [hunk 1]
```diff
@@ -40,8 +35,8 @@ extern "C" {
 /**
  * @def ENABLE_DEBUG
  * @brief   This macro can be defined as 0 or other on a file-based level.
- *          If ENABLE_DEBUG is 0 @ref DEBUG() and @ref DEBUGF() will generate
- *          no output if not they will generate output.
+ *          If ENABLE_DEBUG is 0 @ref DEBUG() will generate no output if not
+ *          they will generate output.
```
Comment (kaspar030) [inline core/include/debug.h]: s/they/it/ ?

File: sys/crypto/3des.c [hunk 2]
```diff
@@ -326,8 +327,8 @@ int tripledes_decrypt(const cipher_context_t *context, const uint8_t *crypt, uin
     res = des3_key_setup(context->context, key);
 
     if (res < 0) {
-        DEBUGF("[ERROR] des3_key_setup failed with Code %i\r\n", res);
-        free(key);
+        DEBUG("%s:%d in %s: [ERROR] des3_key_setup failed with Code %i\r\n",
+              __FILE__, __LINE__, DEBUG_FUNC, res); free(key);
```
Comment (kaspar030) [inline sys/crypto/3des.c]: newline

**Comment by kaspar030** [comment]
IMO you should use RIOT_FILE_RELATIVE instead of __FILE__.

File: core/include/debug.h [hunk 3]
```diff
@@ -100,22 +95,15 @@ extern "C" {
...
-#define DEBUGF(...) \
-    do { \
-        DEBUG_PRINT("DEBUG(%s): %s:%d in %s: ", \
-                sched_active_thread ? sched_active_thread->name : "NO THREAD", \
-                __FILE__, __LINE__, DEBUG_FUNC); \
-        DEBUG_PRINT(__VA_ARGS__); \
-    } while (0)
+
+#if DEVELHELP
+#   define THREAD_NAME      (sched_active_thread->name)
```
Comment (gebart) [inline core/include/debug.h]: Is this macro used anywhere now that DEBUGF was removed?

File: core/include/debug.h (hunk 1 above)
Comment (gebart) [inline core/include/debug.h]: How about:
"`DEBUG()` will generate output only if `ENABLE_DEBUG` is non-zero" (or `!= 0`)

**Comment by OlegHahm** [comment]
addressed comments

**Comment by OlegHahm** [comment]
@kaspar030, are you fine?

File: cpu/msp430-common/include/assert.h [hunk 4]
```diff
@@ -51,8 +51,8 @@ extern "C" {
 #ifndef NDEBUG
 # define assert(assertion) \
     if (!(assertion)) { \
-        printf("%s:%d: %s: Assertion '%s' failed", __FILE__, __LINE__, \
-               ASSERT_FUNC, #assertion); \
+        printf("%s:%d: %s: Assertion '%s' failed", RIOT_FILE_RELATIVE,
```
Comment (gebart) [inline cpu/msp430-common/include/assert.h]: missing backslash at end of line

**Comment by gebart** [comment]
I'm fine with dropping DEBUGF and the changes in general, need to see what Travis says after the missign backslash is added though.

File: sys/include/embUnit/AssertImpl.h [hunk 5]
```diff
@@ -59,9 +59,9 @@ void assertImplementationCStr(const char *expected,const char *actual, long line
         long long ____expected__ = (long long) (expected_); \
         long long ____actual__ = (long long) (actual_); \
         if (____expected__ != ____actual__) { \
-            assertImplementationLongLong(____expected__, ____actual__, __LINE__, __FILE__); \
-            return; \
-        } \
+            assertImplementationLongLong(____expected__, ____actual__, \
+                                         __LINE__, RIOT_FILE_RELATIVE); \
+                return; \ } \
```
Comment (cgundogan) [inline sys/include/embUnit/AssertImpl.h]: indent seems to be off

**Comment by OlegHahm** [comment]
Addressed @gebart's and @cgundogan's comments. Ready to squash?

**Comment by kaspar030** [comment]
please squash!

**Comment by OlegHahm** [comment]
Squashed with emtpy diff to the previous state.

**Comment by OlegHahm** [comment]
rebased

**Comment by OlegHahm** [comment]
@gebart, @cgundogan, aynone else: ACK?

**Comment by cgundogan** [comment]
looks sane. I ACK

File: sys/include/embUnit/AssertImpl.h [hunk 6]
```diff
@@ -49,7 +49,7 @@ void assertImplementationCStr(const char *expected,const char *actual, long line
         __typeof__(expected_) ____expected__ = expected_; \
         __typeof__(actual_) ____actual__ = actual_; \
         if (stdimpl_strcmp(____expected__, ____actual__) != 0) { \
-            assertImplementationCStr(____expected__, ____actual__, __LINE__, __FILE__); \
+            assertImplementationCStr(____expected__, ____actual__, __LINE__, RIOT_FILE_RELATIVE); \
```
Comment (cgundogan) [inline sys/include/embUnit/AssertImpl.h]: Should we really change this here? When someone acts upon the great idea to upgrade to another version of embUnit, this change my get lost?

Comment (OlegHahm) [inline sys/include/embUnit/AssertImpl.h]: This is clearly a problem and fault of this strange future person. ;)

**Comment by OlegHahm** [comment]
Need a second ACK.

File: sys/include/embUnit/AssertImpl.h (hunk 6 above)
Comment (authmillenon) [inline sys/include/embUnit/AssertImpl.h]: The last version of [embunit is from 2004](http://sourceforge.net/projects/embunit/files/?source=navbar), so an update would be a real surprise.

**Comment by authmillenon** [comment]
This PR does a lot more then it title states oO

File: cpu/msp430-common/include/assert.h [hunk 7]
```diff
@@ -51,8 +51,8 @@ extern "C" {
 #ifndef NDEBUG
 # define assert(assertion) \
     if (!(assertion)) { \
-        printf("%s:%d: %s: Assertion '%s' failed", __FILE__, __LINE__, \
-               ASSERT_FUNC, #assertion); \
+        printf("%s:%d: %s: Assertion '%s' failed", RIOT_FILE_RELATIVE, \
+                __LINE__, \ ASSERT_FUNC, #assertion); \
```
Comment (authmillenon) [inline cpu/msp430-common/include/assert.h]: What's the backslash doing in there?

Comment (OlegHahm) [inline cpu/msp430-common/include/assert.h]: oops

**Comment by OlegHahm** [comment]
> This PR does a lot more then it title states oO

Tell the reviewers! ;-)

Updated the title.

File: cpu/msp430-common/include/assert.h (hunk 7 above)
Comment (OlegHahm) [inline cpu/msp430-common/include/assert.h]: fixed

**Comment by authmillenon** [comment]
ACK, please squash

**Comment by OlegHahm** [comment]
squashed

**Comment by authmillenon** [comment]
Travis is not happy anymore :(

**Comment by OlegHahm** [comment]
This is unrelated to this PR. Let me create a fix PR.

**Comment by authmillenon** [comment]
Which one?

**Comment by OlegHahm** [comment]
Sorry, this one: https://github.com/RIOT-OS/RIOT/pull/3899

**Comment by OlegHahm** [comment]
Rebased to master.

@authmillenon and @cgundogan, do your ACKs hold?

**Comment by authmillenon** [comment]
Mine does. Did you have conflicts while rebasing?

**Comment by OlegHahm** [comment]
Nope.

**Comment by authmillenon** [comment]
Than it's save to assume that your code did not change. :-) Re-ACK (officially)

**Comment by gebart** [comment]
ACK

**Comment by OlegHahm** [comment]
thx for the review

**Pull request merged**